Moving E/gamma ID variables to the reco::Photon/reco::GsfElectron; update HEEP ID, update IDs  & stored for MINIAOD

Sam-Harper opened the pull request:
~~This is a WIP, there are a few outstanding bugs I'm trying to pin down.  I could also split this PR into two, most of the changes is adapting VID to read from the object.  It will not pass unit tests (because of tau problem which generical to all cms) nor matrix (one test still to fix).~~: Now fixed

## PR description:

In E/gamma IDs, variables are sometimes calculated on “on the fly” during the computation of the ID. While we have survived with this, it would be much nicer if the variables were in the electron/photon as this means we dont have to continuously re-run expensive producers. 

This PR does two things: it moves wherever possible an ID variable to the the photon/electron object and ensures the correct ID information is propagated to the miniAOD. At the same time it adds in HEEPV7.1 ID which is a bug fix ID which changes the pt cut to 5 GeV and uses calo energy to scale the isolation cut. 

A mechanism to update <105X samples exists via RecoEgamma/EgammaTools/plugins/EG9X105XObjectUpdateModifier.cc

### Photon Variable changes

#### particle flow isolation:

it turns out only pfcluster isolation, charged, neutral and photon were being filled so the others were removed 

photon and neutral isolation use the isolation calculated by PhotonIDValueMap producer as that is what is used in E/gamma cut based and MVA IDs. These two variables are identifical to the existing photon & neutral isolations.  

charged isolation: this is a practical but not fundamental definition change. The existing charged isolation uses the particle flow vtx association to determine vertex. The current charged isolation uses dxy, dz cuts instead. While the two isolations will disagree, they are fundamentally the same thing so the new variable can replace the existing one. Again this is the variable photon IDs actually use.  The old charged isolation is now chargedHadronPFPVIso incase anybody really wants it. 

chargedHadronWorstVtxIso and chargedHadronWorstVtxGeomVetoIso have been added

#### Other ids: showershape:
smajor, sminor, smajorAlpha used in long lived analyses have been added.  There is a bug long standing bug in the code to calculate them, it doesnt check it actually finds a rec-hit and blindly uses it anyways. Patch incoming for that. 

### Electron Variable changes:

the HEEP track isolation is added to the object at the AOD level (it was already in the miniAOD). I would have liked to simply replace the tracker isolation variable but it changes the PF event content (its used in the ID). There is no reason to suspect that the changes are worse (or better) but at this point its best not to touch it. 

conversion vtx probablity: this is now added to the electron as its used in the MVA. This causing me major problems trying to get it to agree, this variable is very buggy and all the more reason to be in the object!

helperfuncs:  closestTrkNHits() and closetTrkChi2() allows these functions to be more easily used in our framework

The change to the dataformats is just a few files, most of the changes come from VID and adapting it to use the new format. 

### Expected physics changes:
heep trk isolation (and thus ID) will slight change on some events, this is because we’ve gone from calculating on PackedCandidates to actual tracks so we encounter some small differences due to precision. 
photon : photon & neutral hadron isolations will in rare cases change slightly, due to a small difference in vetos between CITK and PhotonID isolation computations. The variables remain fundementally the same though
photon : charged isolation is now a different but conceptually similar variable, the old one remains
IDs : no ID or MVA variable changes should happen except  a few events for the HEEP IDs changing due to rounding errors. 


## PR validation:

An event by event comparison on 2018 EGamma data has been performed. Its not quite closing yet but soon (mainly as I'm finding lots of existing bugs...)

## To Do:

1. ~~problems with calculating the conversion vertex fit probability and some disagreements in the photon MVA.~~
1. ~~3 problems in matrix tests, I know how to fix two of them (its a bug elsewhere), the 3rd I need to resolve~~
1. ~~photon MVA isnt agreement well, something is bugged there~~ now very minor

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9141

- This PR adds an extra 228KB to repository

cmsbuild commented:
A new Pull Request was created by  @Sam-Harper (Sam Harper) for master.

It involves the following packages:

DataFormats/EgammaCandidates
PhysicsTools/NanoAOD
PhysicsTools/PatAlgos
RecoEgamma/EgammaElectronAlgos
RecoEgamma/EgammaElectronProducers
RecoEgamma/EgammaPhotonProducers
RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification
RecoEgamma/PhotonIdentification
RecoParticleFlow/PFProducer


@perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @jainshilpi, @hatakeyamak, @rappoccio, @HeinerTholen, @varuns23, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @acaudron, @lgray, @jdolen, @drkovalskyi, @ferencek, @rovere, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @clelange, @JyothsnaKomaragiri, @mverzett, @cbernet, @gpetruc, @mariadalfonso, @pvmulder, @bachtis this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: d5a9171d81fe8fd8335b657b7d255d943191f863

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34058/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>20034.0 step3
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17.log

</pre>21234.0 step3
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D21_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D21+RecoFullGlobal_2023D21+HARVESTFullGlobal_2023D21.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

Sam-Harper commented:
Yep as I mentioned, I know the solution for two of those crashes which is improper validity checking of hits found in the cluster2ndMoments. The last (the HION), I'll figure out tomorrow. Additionally I've figured out my discrepancies in the conversion fit probability (its a nasty bug which has been around forever) and the photon MVA problem is just the CITK vs PhotonIDValueMap producer defination of photon / neutral isolation (which should be the same but are not). I'm going to spend an hour or so debugging the photon / neutral isolation bug tomorrow to see what is going on. 

The fall back practical solution is that for the conversion fit vtx prob and photon pf isolation values, I'm going to make a "bugged" data member which stores the delta of the bugged value (which is in use!) with the correct value. It'll be zero most of the time so will compress well.

peruzzim commented on PhysicsTools/NanoAOD/python/electrons_cff.py:
```diff
@@ -341,7 +340,7 @@ def _get_bitmapVIDForEle_docstring(modules,WorkingPoints):
         jetRelIso = Var("?userCand('jetForLepJetVar').isNonnull()?(1./userFloat('ptRatio'))-1.:userFloat('PFIsoAll04')/pt",float,doc="Relative isolation in matched jet (1/ptRatio-1, pfRelIso04_all if no matched jet)",precision=8),
         dr03TkSumPt = Var("?pt>35?dr03TkSumPt():0",float,doc="Non-PF track isolation within a delta R cone of 0.3 with electron pt > 35 GeV",precision=8),
-        dr03TkSumPtHEEP = Var("?pt>35?userFloat('dr03TkSumPtHEEP'):0",float,doc="Non-PF track isolation within a delta R cone of 0.3 with electron pt > 35 GeV used in HEEP ID",precision=8),
+        dr03TkSumPtHEEP = Var("?pt>35?dr03TkSumPtHEEP():0",float,doc="Non-PF track isolation within a delta R cone of 0.3 with electron pt > 35 GeV used in HEEP ID",precision=8),
```
isn't this going to give a wrong result when reading older miniAOD releases?

Sam-Harper replied:
the older miniAOD releases will use a modifier to populate their entries. I'll get it running on the nano for the PR, thanks for the catch.

Sam-Harper committed:
updating EGamma dataformats to put missing ID variables in to the object

Sam-Harper committed:
adapting VID/miniAOD/nanoAOD to new format

Sam-Harper committed:
protecting cluster2ndMoments against not found hits

cmsbuild commented:
The code-checks are being triggered in jenkins.

Sam-Harper commented:
so I have fixed my issues. I still have a rare bug giving not identical photon MVA scores but its close enough. 

I also include a fix for the EcalClusterTools::cluster2ndMomentum calculation as it was not checking that it actually found the hit and also removed the taus test which no longer works. 

The conversion fit probablity was found to be a bug in E/gamma ConversionTools (and its semi serious). The issue is that it matches by reference to the conversion tracks (ie product key & index). The probably is that reducedEgamma puts the GsfTracks into a new collection which means the electrons ref is now out of date and therefore the GsfTrack is excluded from being able to match a conversion. This is now sadly baked into E/gamma IDs so thats the new defination.

I'm debating putting a variable (probably as the delta) for proper conversion match.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9170

- This PR adds an extra 188KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

Sam-Harper commented:
now I need to handle the nanoAOD reading different releases however there is another issue https://github.com/cms-nanoAOD/cmssw/pull/308#issuecomment-481187073 which also needs to be resolved and they will happen at the same time

guitargeek commented:
Hi Sam, nice PR and thank you also for the summary talk this morning! Did you already study how much the conversion bug affects the ID? It's true that it's a priori safer to stick with the definitions the ID was trained with, but maybe we should at least check? Because if it would be equal or better, is there a reason to keep the bugged version in the object? Then the version with the delta can be avoided.

Sam-Harper commented:
Hi Jonas,  so no, I dont know exactly. Its completely unclear to me, its certainly something to study.  This also impacts the cut based ID in a slightly different way.  Time is a bit short, so  I would really like to keep the bugged variable around as its safer. Also there is the constraint that in 10_6 we need to be able to read 10_2 data and the IDs when re-run should give exactly the same value. And if we put in just the fixed value, we would have to put the bugged value in its place for 10_2 made samples so the same function would return different variables depending on input sample which to me violates the principle of least surprise and so should be avoided.

Btw, I've been wading through a lot of crusty e/gamma photon code due to this PR, it was a breath of fresh air when I got to your MVA framework. Also thanks for putting a debugging mode in it, it made it trivial for me to find the MVA was changing due to the conversion vtx prob being different.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

peruzzim commented:
> Also there is the constraint that in 10_6 we need to be able to read 10_2 data and the IDs when re-run should give exactly the same value. And if we put in just the fixed value, we would have to put the bugged value in its place for 10_2 made samples so the same function would return different variables depending on input sample which to me violates the principle of least surprise and so should be avoided.

I completely agree - it should behave as if the definition of the input variables (i.e. recipe to get them from MINIAOD) changes according to the version of the ID, not to the version used to produce the input sample or to run on top of it.

cmsbuild commented:
-1

Tested at: 5f54552b362c8bf312f8c84ba00f874bd275c94e

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34122/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testTauEmbeddingProducers had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
The code-checks are being triggered in jenkins.

Sam-Harper commented:
okay I hadnt properly removed the tau test. Its gone fully now

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9204

- This PR adds an extra 232KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @civanch, @mdhildreth, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34122/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1517 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 132
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3140166
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Sam-Harper commented:
So these tests were useful and highlighted a small feature I noticed in my testing too but hadnt fully realised what it meant. Its to do with the selection for a photon going to miniAOD (and what miniAOD content it has) depending on chargedHadronIso. 

Thinking about what to do. I'm in favour of leaving as is and pushing to validation. Reasoning: the new chargedHadronIso isnt fundamentally different, just practically different so it doesnt represent a conceptual change.  The fact that a few photons now pass and some now dont doesnt bother me much.  And I think the new behaviour is now what we want, before we had a cut on chargedIso(old) in the reco and chargedIso(new) in the ID/MVAs. Now we always cut on the same value of chargedIso which is less confusing. So in fact I consider this a bug fix. 

There exists an issue about what to do for backwards compatibility with 102X/94X samples being read in 106X+. This again I think is okay, reading the miniAOD, this is a non-issue. In a re-miniAOD, as it stands, it would change the electrons and photons going into the miniAOD, something I think is okay. We could even set it up so this doesnt happen, will have a think and consult with the rest of the group on the way forward here. However this shouldnt hold up this PR.

slava77 commented:
> At the same time it adds in HEEPV7.1 ID

It would be nice to put a few words to indicate this in the PR title. How about add ```; update HEEP ID```

slava77 commented on RecoParticleFlow/PFProducer/python/particleFlowEGamma_cff.py:
```diff
@@ -53,10 +53,30 @@
 pfNoPileUpIsoSequence*cms.ignore(pfNoPileUpCandidates)*cms.ignore(pfPileUpAllChargedParticles)*\
-egmPhotonIsolationCITK*egmElectronIsolationCITK*egmElectronIsolationPileUpCITK*gedPhotonSequence*gedElectronPFIsoSequence)
+egmPhotonIsolationCITK*egmElectronIsolationCITK*egmElectronIsolationPileUpCITK*\
+photonIDValueMapProducerRECO*\
```
I think that the endline escapes are not needed

Sam-Harper replied:
indeed, they are not, removed

slava77 commented on TauAnalysis/MCEmbeddingTools/test/BuildFile.xml:
```diff
@@ -1,4 +0,0 @@
-<bin name="testTauEmbeddingProducers" file="TestDriver.cpp">
```
https://github.com/cms-sw/cmssw/issues/26385#issuecomment-482567459

Sam-Harper replied:
okay I'll revert this. It was also due to a misunderstanding on units my end, I thought if a single one failed, it would abort things like the comparisions but now I see this is not the case. So this unit test failing is not necessarily a problem for this PR.

slava77 reviewed (commented):
There are some changes in data products and in the code used.
@Sam-Harper please provide a summary of file size changes and the CPU cost differences in the directly affected modules.

It seems from the code like the effects are small.

Sam-Harper commented on DataFormats/EgammaCandidates/interface/Photon.h:
```diff
@@ -460,44 +466,37 @@ namespace reco {
-      float photonIso ;       //  equivalent to sumPhotonPt in  DataFormats/MuonReco/interface/MuonPFIsolation.h
-      float modFrixione ;      
-      float sumChargedParticlePt; //!< sum-pt of charged Particles(inludes e/mu) 
-      float sumNeutralHadronEtHighThreshold;  //!< sum pt of neutral hadrons with a higher threshold
```
They were added in #1174 (https://github.com/cms-sw/cmssw/commit/cbdef671813076b486632fd23a18c998ca4978a0#diff-880776fb5e7897f84b521148d41b3d61) in 2013 (except for modFrixione which was introduced in 2011: https://github.com/cms-sw/cmssw/commit/92dc1cb010059ee30f6ee9a2d666db04b1f31392). 

I have checked an 80X produced file and they are all zero there. I can not exclude the possibility that RunI data fills modFrixione but searching through the code as much as I could, I think the answer is no.

Sam-Harper commented on DataFormats/EgammaCandidates/interface/Photon.h:
```diff
@@ -166,6 +166,9 @@ namespace reco {
       float e2x5Bottom;
+      float smajor;
+      float sminor;
+      float smajorAlpha;
```
so in e/gamma we've always called them this (well smajorAlpha to be just alpha) but I agree with you, this name is better :)  I asked around the rest of E/gamma management and they have no objections. 

done (and thanks!)

Sam-Harper commented on DataFormats/EgammaCandidates/interface/Photon.h:
```diff
@@ -460,44 +466,37 @@ namespace reco {
-      float sumChargedParticlePt; //!< sum-pt of charged Particles(inludes e/mu) 
-      float sumNeutralHadronEtHighThreshold;  //!< sum pt of neutral hadrons with a higher threshold
-      float sumPhotonEtHighThreshold;  //!< sum pt of PF photons with a higher threshold
-      float sumPUPt;  //!< sum pt of charged Particles not from PV  (for Pu corrections)
```
none that I'm aware of. At this point we are limiting the AOD object isolation to isolation used in central e/gamma IDs.

Sam-Harper commented on DataFormats/EgammaCandidates/src/classes_def.xml:
```diff
@@ -53,10 +54,14 @@
    <version ClassVersion="12" checksum="114160170"/>
    <version ClassVersion="13" checksum="3959509359"/>
+   <version ClassVersion="14" checksum="4260050122"/>
+   <ioread sourceClass="reco::Photon::PflowIsolationVariables" version="[1-13]" source="float chargedHadronIso" targetClass="reco::Photon::PflowIsolationVariables" target="chargedHadronPFPVIso">
```
done

Sam-Harper commented on RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc:
```diff
@@ -734,6 +740,13 @@ void GsfElectronAlgo::createElectron(reco::GsfElectronCollection & electrons, El
+  //relinked to a new collection which means it can no longer match the conversion
+  //as it matches based on product/id
+  //we keep this defination for the MVAs
+  conversionVars.vtxFitProb = ConversionTools::getVtxFitProb(ConversionTools::matchedConversion(electronData.coreRef->ctfTrack(), *eventData.conversions, eventData.beamspot->position(),2.0,1e-6,0));
```
done

Sam-Harper commented on RecoParticleFlow/PFProducer/python/particleFlowEGamma_cff.py:
```diff
@@ -53,10 +53,30 @@
 egmElectronIsolationPileUpCITK.srcToIsolate = cms.InputTag("gedGsfElectronsTmp")
 egmElectronIsolationPileUpCITK.srcForIsolationCone = cms.InputTag("pfPileUpAllChargedParticles")
 
+photonIDValueMapProducerRECO = cms.EDProducer(
```
so photonIDValueMapProducer is removed but the ideal was I wanted to keep the logical connection to it while giving it a new name. This is infact important incase somebody is still using it in their private code. 

However if you really want ,  I can change it, I just thought it was clearer this way (for the short term at least, long term you are probably right).

slava77 replied:
IIUC, these configs can only be done in reco. 
Is there a common alternative version that does the same in miniAOD to justify the suffix?
I prefer that "RECO" is not added.

Sam-Harper replied:
well there was. Anyways, I see your point, changed to photonIDValueMaps as suggested and will push shortly

Sam-Harper commented on RecoEgamma/EgammaPhotonProducers/python/ootPhotons_cff.py:
```diff
@@ -15,9 +15,14 @@
+        photonIso = cms.InputTag(""),
+        chargedHadronWorstVtxIso = cms.InputTag(""),
+        chargedHadronWorstVtxGeomVetoIso = cms.InputTag(""),
+        chargedHadronPFPVIso = cms.InputTag(""),
```
done

Sam-Harper replied:
so undone,  turns out pf isolation cfg doesnt exist in the default gedPhotons config. I know, this is awful, the GEDPhoton producer is a festering mess and its in my crosshairs for 11_X.

Sam-Harper commented on RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:
```diff
@@ -80,12 +80,19 @@ GEDPhotonProducer::GEDPhotonProducer(const edm::ParameterSet& config) :
-    phoPhotonIsolationTokenCITK_      = 
-      consumes<edm::ValueMap<float>>(conf_.getParameter<edm::InputTag>("photonIsolation"));   
+    const edm::ParameterSet& pfIsolCfg = conf_.getParameter<edm::ParameterSet>("pfIsolCfg");
+    auto getVMToken = [&pfIsolCfg,this](const std::string& name){
```
the consumes call is a member function of the producer, hence the "this".

Sam-Harper commented on RecoEgamma/EgammaPhotonProducers/src/GEDPhotonProducer.cc:
```diff
@@ -80,12 +80,19 @@ GEDPhotonProducer::GEDPhotonProducer(const edm::ParameterSet& config) :
+    phoChargedPFPVIsoToken_ = getVMToken("chargedHadronPFPVIso");
+
+
+
```
done

Sam-Harper commented on RecoEgamma/EgammaTools/plugins/EG9X105XObjectUpdateModifier.cc:
```diff
@@ -0,0 +1,175 @@
+  ~EG9X105XObjectUpdateModifier() override{}
+  
+  void setEvent(const edm::Event&) final;
+  void setEventContent(const edm::EventSetup&) final;
```
Fully agree. However this is what it was misnamed as in the base class:
https://github.com/cms-sw/cmssw/blob/master/CommonTools/CandAlgos/interface/ModifyObjectValueBase.h#L26

so i'm kinda stuck with it.

Sam-Harper commented on RecoEgamma/EgammaTools/plugins/EG9X105XObjectUpdateModifier.cc:
```diff
@@ -0,0 +1,175 @@
+class EG9X105XObjectUpdateModifier : public ModifyObjectValueBase {
+public:
+  template<typename T>
+  struct TokenHandlePair{
```
done

Sam-Harper commented on RecoEgamma/EgammaTools/plugins/EG9X105XObjectUpdateModifier.cc:
```diff
@@ -0,0 +1,175 @@
+  
+  reco::GsfElectron::IsolationVariables isolVars03 = ele.dr03IsolationVariables();
+  isolVars03.tkSumPtHEEP = (*eleTrkIso_.handle)[ptrForVM];
+  ele.setDr03Isolation(isolVars03);
```
yes, but unfortunately that variable isnt calculated by the map. I do understand that not filling it is a bit of a mild problem. No action right now but will try and fix.

slava77 replied:
is it calculated with the new setup now?
There access method is added/present.

Sam-Harper replied:
so now the HEEP ID value map producer calculates cone of 0.4. Its never going to be used though :)

This whole thing  comes from the support of cones of 0.3 and 0.4 which will be reviewed in the future. 0.4 I believe is used by nobody these days.

Sam-Harper commented on RecoEgamma/EgammaTools/python/egammaObjectModificationsInMiniAOD_cff.py:
```diff
@@ -5,12 +5,15 @@
 #photon mva ids
 import RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring16_nonTrig_V1_cff as pho_spring16_nt_v1
-import RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Fall17_94X_V1_cff as pho_fall17_94X_v1
+#import RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Fall17_94X_V1_cff as pho_fall17_94X_v1
```
agreed, oversight, thanks

Sam-Harper commented on RecoEgamma/EgammaTools/src/ConversionTools.cc:
```diff
@@ -67,6 +67,24 @@ bool ConversionTools::matchesConversion(const reco::GsfElectron &ele, const reco
+  //closest ctf track ref
+
+  const std::vector<edm::RefToBase<reco::Track> > &convTracks = conv.tracks();
+  for (std::vector<edm::RefToBase<reco::Track> >::const_iterator it=convTracks.begin(); it!=convTracks.end(); ++it) {
```
done

Sam-Harper commented on RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleEmHadD1IsoRhoCut.cc:
```diff
@@ -61,9 +61,11 @@ GsfEleEmHadD1IsoRhoCut::
-  const float isolEmHadDepth1 = cand->dr03EcalRecHitSumEt() + cand->dr03HcalDepth1TowerSumEt(); 
+  const float isolEmHadDepth1 = cand->dr03EcalRecHitSumEt() + cand->dr03HcalDepth1TowerSumEt();
+
+  const float sinTheta = cand->energy()!=0. ? cand->et()/cand->energy() : 0.;
```
indeed, my electron bias showing. done

Sam-Harper commented on RecoEgamma/ElectronIdentification/plugins/cuts/GsfEleMinEcalEtCut.cc:
```diff
@@ -0,0 +1,46 @@
+GsfEleMinEcalEtCut::
+operator()(const reco::GsfElectronPtr& cand) const{  
+  const reco::GsfElectron& ele = *cand;
+  const float sinTheta = ele.energy()!=0 ? ele.et()/ele.energy() : 0.;
```
done

Sam-Harper commented on RecoEgamma/EgammaTools/interface/EleEnergyRetriever.h:
```diff
@@ -0,0 +1,39 @@
+    else if(type_==EnergyType::SuperCluster) return ele.superCluster()->energy();
+    else if(type_==EnergyType::SuperClusterRaw) return ele.superCluster()->rawEnergy();
+    else {
+      throw cms::Exception("LogicError") <<" function has not been updated for "<<static_cast<int>(type_)<<" please do so";
```
ah I didnt know that, good call. Done

Sam-Harper commented on RecoEgamma/ElectronIdentification/python/Identification/heepElectronID_HEEPV71_cff.py:
```diff
@@ -0,0 +1,95 @@
+
+WP_HEEP71_EE = HEEP_WorkingPoint_V1(
+     idName=idName,     
+    dEtaInSeedCut=0.006,     
```
fixed

Sam-Harper commented on RecoEgamma/PhotonIdentification/python/Identification/cutBasedPhotonID_tools.py:
```diff
@@ -171,15 +170,16 @@ def psetChHadIsoWithEALinScalingCut(wpEB, wpEE, isoInputs):
+        cutName = cms.string('PhoGenericRhoPtScaledCut'), 
+        cutVariable = cms.string("chargedHadronIso"),
+        lessThan = cms.bool(True),
+        # cut 
```
done, its now part of the function comment

Sam-Harper commented on RecoEgamma/PhotonIdentification/python/PhotonMVAValueMapProducer_cfi.py:
```diff
@@ -8,8 +8,8 @@
-from RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Fall17_94X_V1_cff import *
-mvaConfigsForPhoProducer.append( mvaPhoID_RunIIFall17_v1_producer_config )
+#from RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Fall17_94X_V1_cff import *
+#mvaConfigsForPhoProducer.append( mvaPhoID_RunIIFall17_v1_producer_config )
```
removed

Sam-Harper commented:
> There are some changes in data products and in the code used.
> @Sam-Harper please provide a summary of file size changes and the CPU cost differences in the directly affected modules.

This is in progress and expect a complete set of results this evening. It is not expected to be large, initial tests show 0.4% increase for slimmedElectrons and 1% for slimmedPhotons running on EGamma 2018 dataset with the total effect on the file size as reported by "ls" being no change.

Sam-Harper committed:
code review comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9279

- This PR adds an extra 488KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

Sam-Harper commented:
Okay first results of timing and diskspace in:
data samples: 
EGamma 2018 , 316058, LS 60,  1K events 
JetHT 2018: 322252, LS 80, 1K events

size on disk (bytes)

| sample | ref |  new |  change  |  
| ---| --- |---|---|
| Egamma AOD |  431865103  |  432187368 | +0.07% |
| Egamma miniAOD | 58896326 | 58861553 | -0.06% |
| JetHT AOD | 410829202 |  411015969 | 0.04% |
| JetHT miniAOD | 55616816 | 55586650 | -0.06% | 

From jetHT:
patElectrons_slimmedElectrons__reRECO. 3047.91 821.307
patPhotons_slimmedPhotons__reRECO. 1000.13 231.005
to 
patElectrons_slimmedElectrons__reRECO. 3079.88 820.549
patPhotons_slimmedPhotons__reRECO. 909.728 222.536

From e/gamma:
patElectrons_slimmedElectrons__reRECO. 7400.65 1901.78
patPhotons_slimmedPhotons__reRECO. 4785.06 1168.24
to
patElectrons_slimmedElectrons__reRECO. 7480.99 1903.83
patPhotons_slimmedPhotons__reRECO. 4356.82 1120.37

For e/gamma AOD:
recoPhotons_gedPhotons__reRECO. 2618.28 1885.62
recoGsfElectrons_gedGsfElectrons__reRECO. 2646 1785.37
to
recoPhotons_gedPhotons__reRECO. 2669.03 1984.21
recoGsfElectrons_gedGsfElectrons__reRECO. 2686.36 1808.91

This is as expected, we reduce the user data for pat::Electrons/Photons so miniAOD works out marginally smaller. AOD we only add info so its a bit bigger but AOD is so large w.r.t the e/gamma collections there is no impact.

Timing: E/gamma (RECO + MINIAOD job)

| delta/mean | delta/orJob | original | new | name  | 
| ---| --- |---|---|---|
| removed | -0.01% | 1.55 ms/ev | 0.00 ms/ev | heepIDVarValueMaps |
| removed | -0.05% | 8.23 ms/ev | 0.00 ms/ev | photonIDValueMapProducer |
| added | +0.09% | 0.00 ms/ev | 13.68 ms/ev | photonIDValueMapProducerRECO |
| removed | -0.00% | 0.25 ms/ev | 0.00 ms/ev | egmPhotonIsolation |
| -1.808031 | -0.01% | 0.97 ms/ev | 0.05 ms/ev | ctppsDiamondRecHits |
| -1.742736 | -0.02% | 3.32 ms/ev | 0.23 ms/ev | patJetCorrFactorsAK8 |
| -1.662091 | -0.01% | 0.87 ms/ev | 0.08 ms/ev | patJetCorrFactorsAK8Puppi |
| -1.635088 | -0.01% | 1.04 ms/ev | 0.10 ms/ev | gtDigis |
| +1.151515 | +0.00% | 0.01 ms/ev | 0.03 ms/ev | ak4PFCHSResidualCorrector |
| -1.121255 | -0.04% | 8.31 ms/ev | 2.34 ms/ev | dt1DRecHits |
| +0.804424 | +0.03% | 3.41 ms/ev | 7.99 ms/ev | muonCSCDigis |
| -0.723438 | -0.01% | 1.71 ms/ev | 0.80 ms/ev | muonDTDigis |
| -0.561243 | -0.00% | 1.40 ms/ev | 0.79 ms/ev | QGTagger |
| -0.557611 | -0.03% | 10.58 ms/ev | 5.96 ms/ev | csc2DRecHits |
| +0.417690 | +0.00% | 0.97 ms/ev | 1.48 ms/ev | castorreco |
| -0.401671 | -0.00% | 2.01 ms/ev | 1.34 ms/ev | MeasurementTrackerEventPreSplitting |
| +0.388164 | +0.03% | 9.89 ms/ev | 14.65 ms/ev | pfCombinedCvsLJetTags |
| +0.387707 | +0.00% | 1.36 ms/ev | 2.02 ms/ev | ecalRecHit |
| -0.376404 | -0.01% | 2.96 ms/ev | 2.02 ms/ev | gedGsfElectrons |
| -0.337486 | -0.00% | 1.03 ms/ev | 0.73 ms/ev | oniaPhotonCandidates |
| -0.290987 | -0.00% | 0.81 ms/ev | 0.60 ms/ev | photonMVAValueMapProducer |
| +0.288276 | +0.01% | 6.11 ms/ev | 8.17 ms/ev | pfCombinedCvsBJetTags |
| +0.281170 | +0.00% | 1.35 ms/ev | 1.79 ms/ev | dt1DCosmicRecHits |
| +0.247317 | +0.00% | 0.94 ms/ev | 1.20 ms/ev | totemRPUVPatternFinder |
| +0.215008 | +0.01% | 5.03 ms/ev | 6.24 ms/ev | pfImpactParameterTagInfosSlimmedDeepFlavour |
| -0.155100 | -0.01% | 10.45 ms/ev | 8.95 ms/ev | initialStepSeedsPreSplitting |
| +0.154680 | +0.03% | 30.14 ms/ev | 35.19 ms/ev | siPixelClustersPreSplitting |
| +0.144300 | +0.01% | 10.24 ms/ev | 11.84 ms/ev | gedGsfElectronsTmp |
| +0.137533 | +0.02% | 21.20 ms/ev | 24.33 ms/ev | lowPtGsfElectrons |
| +0.132523 | +0.00% | 4.92 ms/ev | 5.62 ms/ev | ecalDrivenGsfElectrons |
| +0.107962 | +0.01% | 7.68 ms/ev | 8.55 ms/ev | siStripZeroSuppression |
| +0.074234 | +0.01% | 11.58 ms/ev | 12.48 ms/ev | siPixelRecHitsPreSplitting |
| -0.066034 | -0.00% | 10.50 ms/ev | 9.83 ms/ev | ecalPreshowerRecHit |
| +0.060958 | +0.00% | 8.35 ms/ev | 8.88 ms/ev | siPixelDigis |

total 15.1542 s/ev  -> 15.3237 s/ev

JetHT (RECO + MINIAOD job)

| delta/mean | delta/orJob | original | new | name  | 
| ---| --- |---|---|---|
| removed | -0.00% | 0.84 ms/ev | 0.00 ms/ev | heepIDVarValueMaps |
| removed | -0.01% | 1.72 ms/ev | 0.00 ms/ev | photonIDValueMapProducer |
| added | +0.03% | 0.00 ms/ev | 4.31 ms/ev | photonIDValueMapProducerRECO |
| removed | -0.00% | 0.07 ms/ev | 0.00 ms/ev | egmPhotonIsolation |
| +1.971119 | +0.01% | 0.01 ms/ev | 1.10 ms/ev | totemRPUVPatternFinder |
| +1.900884 | +1.22% | 5.48 ms/ev | 215.60 ms/ev | csc2DRecHits |
| +1.884514 | +0.00% | 0.01 ms/ev | 0.37 ms/ev | hpsPFTauDiscriminationByMVA6LooseElectronRejectionBoosted |
| -1.875000 | -0.00% | 0.06 ms/ev | 0.00 ms/ev | ak4PFCHSL3AbsoluteCorrectorNoHF |
| -1.724179 | -0.07% | 12.53 ms/ev | 0.93 ms/ev | ctppsDiamondRawToDigi |
| +1.709538 | +1.15% | 16.74 ms/ev | 213.74 ms/ev | l1extraParticles |
| -1.573333 | -0.00% | 0.07 ms/ev | 0.01 ms/ev | ak4PFCHSResidualCorrector |
| -1.562842 | -0.00% | 0.33 ms/ev | 0.04 ms/ev | castorreco |
| -1.536504 | -0.01% | 1.32 ms/ev | 0.17 ms/ev | totemRPRawToDigi |
| +1.520327 | +0.54% | 14.72 ms/ev | 108.02 ms/ev | siStripZeroSuppression |
| +1.479872 | +0.55% | 16.68 ms/ev | 111.62 ms/ev | ecalPreshowerRecHit |
| -1.466667 | -0.00% | 0.01 ms/ev | 0.00 ms/ev | ak4PFPuppiL3AbsoluteCorrector |
| +1.455499 | +0.06% | 1.92 ms/ev | 12.19 ms/ev | ecalRecHit |
| -1.441860 | -0.00% | 0.07 ms/ev | 0.01 ms/ev | ak4PFCHSL2RelativeCorrectorNoHF |
| +1.434316 | +0.01% | 0.21 ms/ev | 1.28 ms/ev | displacedMuonSeeds |
| +1.108128 | +0.00% | 0.30 ms/ev | 1.04 ms/ev | cosmicsVetoTracksRaw |
| +1.062812 | +0.01% | 0.47 ms/ev | 1.54 ms/ev | ctppsPixelClusters |
| +0.976476 | +0.49% | 44.53 ms/ev | 129.50 ms/ev | siPixelClustersPreSplitting |
| +0.964688 | +0.01% | 0.69 ms/ev | 1.97 ms/ev | dt1DCosmicRecHits |
| -0.909688 | -0.06% | 15.95 ms/ev | 5.98 ms/ev | muonCSCDigis |
| -0.693323 | -0.01% | 1.68 ms/ev | 0.82 ms/ev | pfCombinedCvsBJetTagsPuppi |
| -0.647872 | -0.01% | 2.40 ms/ev | 1.22 ms/ev | ctppsPixelLocalTracks |
| +0.633344 | +0.57% | 105.69 ms/ev | 203.65 ms/ev | ecalMultiFitUncalibRecHit |
| -0.629874 | -0.00% | 1.10 ms/ev | 0.57 ms/ev | initialStepSeedLayersPreSplitting |
| +0.560400 | +0.01% | 2.30 ms/ev | 4.10 ms/ev | dt1DRecHits |
| -0.504625 | -0.05% | 19.63 ms/ev | 11.72 ms/ev | siPixelRecHitsPreSplitting |
| -0.497182 | -0.00% | 1.00 ms/ev | 0.60 ms/ev | oniaPhotonCandidates |
| -0.477415 | -0.01% | 3.95 ms/ev | 2.43 ms/ev | pfCombinedCvsLJetTagsPuppi |
| -0.459640 | -0.01% | 3.59 ms/ev | 2.25 ms/ev | MeasurementTrackerEventPreSplitting |
| -0.374311 | -0.00% | 2.48 ms/ev | 1.70 ms/ev | gedGsfElectrons |
| +0.295952 | +0.01% | 6.51 ms/ev | 8.78 ms/ev | pfCombinedCvsBJetTags |
| +0.279777 | +0.01% | 7.24 ms/ev | 9.60 ms/ev | standAloneMuons |
| +0.229283 | +0.02% | 11.87 ms/ev | 14.94 ms/ev | pfCombinedCvsLJetTags |
| +0.195334 | +0.01% | 7.79 ms/ev | 9.48 ms/ev | siPixelDigis |
| +0.138854 | +0.01% | 7.44 ms/ev | 8.55 ms/ev | gedGsfElectronsTmp |
| +0.133869 | +0.02% | 27.18 ms/ev | 31.08 ms/ev | lowPtGsfElectrons |
| +0.107189 | +0.00% | 6.21 ms/ev | 6.91 ms/ev | pfImpactParameterTagInfosSlimmedDeepFlavour |
| -0.096355 | -0.00% | 9.12 ms/ev | 8.28 ms/ev | initialStepSeedsPreSplitting |
| -0.082210 | -0.19% | 420.98 ms/ev | 387.74 ms/ev | tobTecStepTrackCandidates |
| +0.075587 | +0.02% | 35.86 ms/ev | 38.68 ms/ev | candidateVertexArbitratorCvsL |
| +0.068442 | +0.01% | 25.70 ms/ev | 27.52 ms/ev | candidateVertexArbitrator |
| -0.063223 | -0.10% | 291.32 ms/ev | 273.47 ms/ev | initialStepTrackCandidatesPreSplitting |
| -0.059326 | -0.26% | 783.14 ms/ev | 738.02 ms/ev | pixelLessStepTrackCandidates |
| +0.058450 | +0.01% | 33.90 ms/ev | 35.94 ms/ev | initialStepHitQuadrupletsPreSplitting |
| -0.057922 | -0.29% | 876.19 ms/ev | 826.87 ms/ev | jetCoreRegionalStepTrackCandidates |
| -0.056531 | -0.03% | 78.70 ms/ev | 74.38 ms/ev | detachedQuadStepTrackCandidates |
| -0.053005 | -0.16% | 534.54 ms/ev | 506.94 ms/ev | lowPtQuadStepTrackCandidates |
| -0.050711 | -0.15% | 532.72 ms/ev | 506.38 ms/ev | highPtTripletStepTrackCandidates |

event 17.1663 -> 17.6425 s/ev

so timing doesnt look that concerning and this is without the speedup that PhotonIDValueMapProducer already has in another PR.

Sam-Harper commented:
also I think I accepted all comments except https://github.com/cms-sw/cmssw/pull/26389#discussion_r274829814  which I'm happy to do so but just wanted to explain my reasoning before I did.

There is also the issue when reading old samples about the HEEP track isolation in a cone of 0.4 which is so far unaddressed but I think is minor. 

Finally on the to do, I have to properly check that my mechanisms for reading old data are working but that I think can probably wait for another update.

fabiocos commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
The tests are being triggered in jenkins.

peruzzim commented:
> There is also the issue when reading old samples about the HEEP track isolation in a cone of 0.4 which is so far unaddressed but I think is minor.
> 
> Finally on the to do, I have to properly check that my mechanisms for reading old data are working but that I think can probably wait for another update.

Processing data produced in previous releases, and reproducing variables used for ID there, do not look like minor items to me... as this is entering the code base - not just for the UL.
In my opinion these issues should be addressed before merging this PR, not with a later fix.
This is what you meant with "another update", right? Could you please clarify otherwise?

cmsbuild commented:
+1
Tested at: a3730c320cd3edb4444a213ab7b0aa3c24af3133
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34205/summary.html

cmsbuild commented:
Comparison job queued.

Sam-Harper commented:
> > There is also the issue when reading old samples about the HEEP track isolation in a cone of 0.4 which is so far unaddressed but I think is minor.
> > Finally on the to do, I have to properly check that my mechanisms for reading old data are working but that I think can probably wait for another update.
> 
> Processing data produced in previous releases, and reproducing variables used for ID there, do not look like minor items to me... as this is entering the code base - not just for the UL.
> In my opinion these issues should be addressed before merging this PR, not with a later fix.
> This is what you meant with "another update", right? Could you please clarify otherwise?

So I would really like this PR in the UL. It'll take another few days to all the possible testing for the pre 106X data so if we wait for that, then this is not going in pre4. If it can still go in pre5, then I'm fine.  However given the only thing that is missing is some minor compatibles with reading old data which we are fully aware of, then I see no reason to hold it.

Sam-Harper committed:
changing pho id vm name

Sam-Harper committed:
heep trk iso 04 is now filled when updating from older releases

cmsbuild commented:
The code-checks are being triggered in jenkins.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9283

- This PR adds an extra 488KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

Sam-Harper commented:
right the ttbar timing and diskspace tests finished on 500 events with PU. Same story as the data tests. 

Last two of Slava's comments are now addressed.  So I think thats everything.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34205/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1505 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 135
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142451
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

slava77 commented on RecoHI/Configuration/python/Reconstruction_hiPF_cff.py:
```diff
@@ -16,6 +16,7 @@
 gedPhotonsTmp.isolationSumsCalculatorSet.trackProducer = cms.InputTag("hiGeneralTracks")
 gedPhotons.primaryVertexProducer = cms.InputTag("hiSelectedVertex")
 gedPhotons.isolationSumsCalculatorSet.trackProducer = cms.InputTag("hiGeneralTracks")
+photonIDValueMapProducerRECO.vertices = cms.InputTag("hiSelectedVertex")
```
this needs a fix

Sam-Harper replied:
urgh right  you are, sorry, taking shortcuts always ends up bitting me :(

Sam-Harper committed:
fixing hion module for photonIDValueMap name change

cmsbuild commented:
The code-checks are being triggered in jenkins.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9286

- This PR adds an extra 236KB to repository

cmsbuild commented:
-1

Tested at: 81d903adb245bf07f25b14aa292666055561c84a

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34211/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
140.53 step2
<pre>runTheMatrix-results/140.53_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI/step2_RunHI2011+RunHI2011+RECOHID11+HARVESTDHI.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

cmsbuild commented:
+1
Tested at: 4fde27e98a7e871f6a21807244e638bf4b32ab63
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34214/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34214/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1505 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 135
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142451
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

slava77 commented:
> ### Expected physics changes:
> IDs : no ID or MVA variable changes should happen except a few events for the HEEP IDs changing due to rounding errors.

Looking at nanoAOD wf 1325.7 (this runs on /RelValTTbar_13/CMSSW_9_4_5_cand1-94X_mc2017_realistic_v14_PU_RelVal_rmaod-v1/MINIAODSIM)

there are quite a few differences
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_10_6_X_2019-04-15-1100+26389/31249/validateJR/all_OldVSNew_TTbar13nanoEDMin2017wf1325p7/

dr03TkSumPtHEEP appears empty
![all_OldVSNew_TTbar13nanoEDMin2017wf1325p7c_nanoaodFlatTable_electronTable__DQM_obj_floats__dr03TkSumPtHEEP_17](https://user-images.githubusercontent.com/4676718/56214900-91e72300-6013-11e9-998c-4562b8c635b4.png)

mvaFall17V1Iso looks like accepting no signal electrons (there should be some, it's a ttbar sample)
![all_OldVSNew_TTbar13nanoEDMin2017wf1325p7c_nanoaodFlatTable_electronTable__DQM_obj_floats__mvaFall17V1Iso_35](https://user-images.githubusercontent.com/4676718/56214960-b04d1e80-6013-11e9-9065-a2e116381d0f.png)

There are also small diffs even in more basic values: pfRelIso03_chg
![all_OldVSNew_TTbar13nanoEDMin2017wf1325p7c_nanoaodFlatTable_photonTable__DQM_obj_floats__pfRelIso03_chg_254](https://user-images.githubusercontent.com/4676718/56215092-eee2d900-6013-11e9-960c-4c48d7926198.png)

I'm guessing that without a way to restore the existing nano workflow output, this PR can not go in.

Sam-Harper commented:
pfRelIso03_chg is by design and is a  feature. 

I'll have a look at the nanos. However there is a fall back solution. I am perfectly happy with putting the VID/nano changes in at a later date, infact they are in a separate commit for this reason. So I can add my variables that I want and then do changes to VID/nano at a later date.

Sam-Harper commented:
Wait, I hadnt noticed. So this is reading the old data. Yes I need to fix that, as stated. Now lets talk schedule, I really really want this in the UL. If it misses pre4, can we still have this in pre5, again the only issue that I'm aware of is the fact we have problems reading the old miniAOD in this work flow.

Sam-Harper commented:
indeed, its easy enough to do so, thats its entire point.  I'm doing it now. Whats the era modifier for nano in 102X called.

peruzzim commented:
We don't have re-miniAOD in 102X for everything. These are the modifiers currently in use:

run2_miniAOD_80XLegacy -> for 80X miniAOD
run2_nanoAOD_94X2016 -> for 94X re-miniAOD of 2016
run2_nanoAOD_94XMiniAODv1, run2_nanoAOD_94XMiniAODv2 -> for 94X 2017
run2_nanoAOD_102Xv1 -> for 2018

what happens when reading 80X?

Sam-Harper commented:
>what happens when reading 80X?
We use the 80X updator

Sam-Harper commented:
right that should do the trick but havent had time to properly test it.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9303

- This PR adds an extra 492KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

Sam-Harper committed:
fixing for reading older releases for nano

slava77 commented:
@peruzzim 
how do we get this tested in the full nano test setup?
we do not have a full range of checks of the nano varieties in cms-bot/jenkins.

cmsbuild commented:
The code-checks are being triggered in jenkins.

Sam-Harper commented:
urgh, okay my hail mary (which worked for the electrons and I had to leave that second and it was looking like it was working photons) had some photon issues which are now fixed. I dont have a way of comparing nanoAOD edm files so I assume they are the same now, tests will show

Sam-Harper commented:
> @peruzzim
> how do we get this tested in the full nano test setup?
> we do not have a full range of checks of the nano varieties in cms-bot/jenkins.

Indeed we should have this. My plan over the next week or so was to calmly test them all and check for differences as we know theres an unrelated problem which needs to be fixed or at least checked.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9310

- This PR adds an extra 252KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

Sam-Harper commented:
Oh and we should really drop the 92X support for nano, its pointless and makes the code less readable

peruzzim commented:
I can do a manual test in the nanoAOD integration area, when everything looks ready to merge here. This actually reminds me to include more configurations in the matrix tests, will do it. And yes, the 92X cleanup has already happened in the integration repository and will come to the upstream repository soon. Thanks

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7d87d85dac54452e68906c4b1810f0153482365c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34235/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34235/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1506 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3142783
 - DQMHistoTests: Total failures: 123
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3142463
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Sam-Harper commented:
Right I now understand why the MVAs changed. Another bug! Happy days. 

It comes from the fact that while in GsfElectron
 virtual TrackRef closestTrack() const { return core()->ctfTrack() ;}
 virtual TrackRef closestCtfTrackRef() const { return core()->ctfTrack() ; }

appear to be interchangeable, they are in fact not. Because only  closestCtfTrackRef is overriden by pat::Electron to use the embedded track and therefore it fails. Fun.  I have fixed it by making closestTrack() call closestCtfTrackRef.  

For the record, this simple, I know I'll change a few variables in the GsfElectron/Photon, has lead to bugs identified in almost all the variables looked at. Sigh.

Sam-Harper committed:
fixing bug where closestTrack& clostestCtfTrackRef give inconsistent results when running where only miniAOD is present

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9326

- This PR adds an extra 252KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
this is getting to a state of a few too many bugfixes/tunings on top of data content migration.
Can we have a split of this PR:
- one to make all necessary bugfixes
- one to be a mostly technical move of variables

I didn't think this through in full in the context of all code changes in this PR, but the split should be a safer way to move on and not miss things that should not change among the things that are expected to change

Sam-Harper commented:
That I can do.  

However the last bug fix however was simply that I used an improper function (which gives identical results when the AOD is present in the event) which was a ticking time bomb that blew up in my face.

slava77 commented on DataFormats/EgammaCandidates/interface/GsfElectron.h:
```diff
@@ -183,12 +183,15 @@ class GsfElectron : public RecoCandidate
-    virtual TrackRef closestTrack() const { return core()->ctfTrack() ; }
+    //so it turns out closestCtfTrackRef is overriden by pat::Electron but not closestTrack()
+    //this results in surprising behaviour, ie two functions which looked identical 
+    //would give different results so now closestTrack just calls closestCtfTrackRef
```
I'm not sure this comment is appropriate. We should not tune up the base class interface/implementations to the decisions/shortcomings of the derived classes.

Both the ```closestTrack``` and ```closestCtfTrackRef``` methods are first appearing in  the ```GsfElectron```.
It seems like your preference is to have them the same.
If the design is for both to be the same; perhaps better remove one and leave just one to avoid confusion.
Edits in other parts of this PR are suggestive that the ```closestTrack``` should be removed.

Sam-Harper replied:
There is a program of work planned for 11_X to fix this and other things like this, now is not the time.

slava77 replied:
this seems simple enough to be done now; considering that it's fractionally already done

Sam-Harper replied:
I'm more thinking about user code. I would prefer not to make code which compiles with 102X not compile with 106X, particularly when we want to clean this & reco::Photon up in 11_X.

slava77 reviewed (commented):
[perhaps I should've commented in the split off portion of this PR]

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
I somehow thought that the last version was working and didn't restart pending the interim PR review.
While running locally I see that in 136.7722 and its MC variant as well for ```Constructing module: class=VersionedGsfElectronIdProducer label='egmGsfElectronIDs'``` there is an exception message  ```MissingParameter: Parameter 'useHEEPIso' not found.```  This runs ```--era Run2_2016,run2_miniAOD_80XLegacy```

cmsbuild commented:
+1
Tested at: 6374152050465896e50e0e24f3dc1b6fef29a5bf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34288/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34288/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1385 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 103
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211657
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

Sam-Harper committed:
removing clostestTrack method on code review advice

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26389/9360

- This PR adds an extra 260KB to repository

cmsbuild commented:
Pull request #26389 was updated. @perrotta, @slava77, @christopheralanwest, @tocheng, @cmsbuild, @franzoni, @fgolf, @tlampen, @pohsun, @santocch, @peruzzim can you please check and sign again.

Sam-Harper commented:
the update here was a mistake, I'm going to close this PR which is superseeded by

https://github.com/cms-sw/cmssw/pull/26485
https://github.com/cms-sw/cmssw/pull/26486


> I somehow thought that the last version was working and didn't restart pending the interim PR review.
> While running locally I see that in 136.7722 and its MC variant as well for `Constructing module: class=VersionedGsfElectronIdProducer label='egmGsfElectronIDs'` there is an exception message `MissingParameter: Parameter 'useHEEPIso' not found.` This runs `--era Run2_2016,run2_miniAOD_80XLegacy`

Interesting, I hadnt seen that. I know the fix, simple. Will be done on the other branch

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 92d523f944c546eca3a2dbf2916f02fa46c33612
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34291/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26389/34291/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1385 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3211964
 - DQMHistoTests: Total failures: 103
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3211657
 - DQMHistoTests: Total skipped: 204
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files
